MovieQuiz/Services/QuestionFactory.swift

```diff
@@ -8,59 +8,114 @@
-        QuizQuestion(
-            image: "Tesla",
-            text: "Рейтинг этого фильма больше чем 6?",
-            correctAnswer: false),
-        QuizQuestion(
-            image: "Vivarium",
-            text: "Рейтинг этого фильма больше чем 6?",
-            correctAnswer: false)
-    ]
-    
-    weak var delegate: QuestionFactoryDelegate?
+    private let moviesLoader: MoviesLoading
+    private let delegate: QuestionFactoryDelegate
     
-    init(delegate: QuestionFactoryDelegate?) {
+    init(moviesLoader: MoviesLoading, delegate: QuestionFactoryDelegate) {
+        self.moviesLoader = moviesLoader
         self.delegate = delegate
     }
     
+    func loadData() {
+        moviesLoader.loadMovies {result in
+            DispatchQueue.main.async { [weak self] in
+                guard let self = self else { return }
+                switch result {
+                case.success(let mostPopularMovies):
+                    self.movies = mostPopularMovies.items
+                    self.delegate.didLoadDataFromServer()
+                case .failure(let error):
+                    self.delegate.didFailToLoadData(with: error)
+                }
+            }
+        }
+    }
+    
     func requestNextQuestion() {
-        let index = (0..<questions.count).randomElement() ?? 0
-        let question = questions[safe: index]
-        delegate?.didReceiveNextQuestion(question: question)
+        DispatchQueue.global().async { [weak self] in
+            guard let self = self else { return }
+            let index = (0..<self.movies.count).randomElement() ?? 0
+            var ratingSet = Set<Float>()
+            for movie in self.movies {
+                ratingSet.insert(Float(movie.rating) ?? 0)
+            }
+
+            guard let movie = self.movies[safe: index] else { return }
+            
+            var imageData = Data()
+            
+            do {
+                imageData = try Data(contentsOf: movie.resizedImageURL)
+            } catch {
+                DispatchQueue.main.async { [weak self] in
+                    let error = error
+                    self?.delegate.didFailToLoadData(with:error)
+                    print("Failed to load image")
+                }
+            }
+            
+            let rating = Float(movie.rating) ?? 0
+            
+            let questionRating: Float = ratingSet.randomElement() ?? 7
+            
+            let text = "Рейтинг этого фильма меньше, чем \(questionRating)?"
+            
+            let correctAnswer = rating < questionRating
+            
+            let question = QuizQuestion(image: imageData,
+                                        text: text,
+                                        correctAnswer: correctAnswer)
+            
+            DispatchQueue.main.async { [weak self] in
+                guard let self = self else { return }
+                self.delegate.didReceiveNextQuestion(question: question)
```

Comment:
В одну строчку же лучше, нет? :) 
```suggestion
                self?.delegate.didReceiveNextQuestion(question: question)
```

Reply:
Оно конечно лучше, но я думал, что guard let обязательно в таких случаях))

Reply:
В данном случае про удобство. Иногда вам нужен доступ к переменным self, чтобы передать переменную в качестве параметра функции self. Тогда чтобы с националами не работать, проще разыменовать и работать с сильной ссылкой. 
Это не ошибка, можно оставит :) скорее дело вкуса, мне такой вариант кажется чище, а кто-то возразит 🙃